### ISSUE TITLE
Xcode warning: Thread running at QOS_CLASS_USER_INTERACTIVE waiting on a lower QoS thread 

### ISSUE THREAD

**Issue by ian-sayles** [firebase/firebase-ios-sdk#10476]
### Description

When uploading a file getting the following purple warning:
GTMSessionFetcherService.m:307 Thread running at QOS_CLASS_USER_INTERACTIVE waiting on a lower QoS thread running at QOS_CLASS_BACKGROUND. Investigate ways to avoid priority inversions

Only started when upgraded Xcode to Version 14.1 (14B47b) and Firebase 10.1.0

<img width="1115" alt="Screenshot 2022-11-10 at 10 51 26" src="https://user-images.githubusercontent.com/4272389/201072129-94ae58ea-30bc-4f0b-8096-d022442ce275.png">

### Firebase SDK Version

10.1.0

### Xcode Version

14.1

### Installation Method

Swift Package Manager

### Firebase Product(s)

Storage

### Targeted Platforms

iOS

### Relevant Log Output

```shell
Domain=_UIViewServiceErrorDomain Code=1 "(null)" UserInfo={Terminated=disconnect method}
Thread Performance Checker: Thread running at QOS_CLASS_USER_INTERACTIVE waiting on a lower QoS thread running at QOS_CLASS_BACKGROUND. Investigate ways to avoid priority inversions
PID: 492, TID: 8412
Backtrace
=================================================================
3  	                          	0x000000010197d850 -[GTMSessionFetcherService sessionForFetcherCreation] + 252
4  	                          	0x00000001019601a4 -[GTMSessionFetcher beginFetchMayDelay:mayAuthorize:mayDecorate:] + 3084
5  	                          	0x0000000101966e78 -[GTMSessionFetcher authorizer:request:finishedWithError:] + 360
6  	                          	0x0000000101966c60 __37-[GTMSessionFetcher authorizeRequest]_block_invoke + 92
7  	                          	0x0000000102d36794 $sSo7NSErrorCSgIeyBy_s5Error_pSgIegg_TR + 120
8  	                          	0x0000000102d3639c $s15FirebaseStorage0B15TokenAuthorizerC16authorizeRequest_17completionHandlerySo19NSMutableURLRequestCSg_ys5Error_pSgctFyycfU1_ + 196
9                             		0x0000000102d27068 $sIeg_IeyB_TR + 48
10  libdispatch.dylib                   0x000000010b0705a8 _dispatch_call_block_and_release + 32
11  libdispatch.dylib                   0x000000010b07205c _dispatch_client_callout + 20
12  libdispatch.dylib                   0x000000010b082810 _dispatch_main_queue_drain + 1196
13  libdispatch.dylib                   0x000000010b082354 _dispatch_main_queue_callback_4CF + 44
14  CoreFoundation                      0x0000000189d2a6f8 5CDC5D9A-E506-3740-B64E-BB30867B4F1B + 632568
15  CoreFoundation                      0x0000000189d0c058 5CDC5D9A-E506-3740-B64E-BB30867B4F1B + 507992
16  CoreFoundation                      0x0000000189d10ed4 CFRunLoopRunSpecific + 612
17  GraphicsServices                    0x00000001c2769368 GSEventRunModal + 164
18  UIKitCore                           0x000000018c1ef3d0 179501B6-0FC2-344A-B969-B4E3961EBE10 + 3810256
19  UIKitCore                           0x000000018c1ef034 UIApplicationMain + 340
20  SwiftUI                             0x000000018d7fc014 494A7D57-90AF-3E90-9623-7038275B8D87 + 1904660
21  SwiftUI                             0x000000018d75d16c 494A7D57-90AF-3E90-9623-7038275B8D87 + 1253740
22  SwiftUI                             0x000000018d7464bc 494A7D57-90AF-3E90-9623-7038275B8D87 + 1160380
23   	                          	0x0000000101551a3c $s10toolminder0A3AppV5$mainyyFZ + 40
24                           		0x0000000101551d30 main + 12
25  dyld                                0x00000001a8368960 CB3FF411-4762-34D2-86A4-ECA13F9FB6C3 + 88416
```

### If using Swift Package Manager, the project's Package.resolved

<details>
<summary>Expand <code>Package.resolved</code> snippet</summary>
<br>

```json

Replace this line with the contents of your Package.resolved.

```

</details>

### If using CocoaPods, the project's Podfile.lock

<details>
<summary>Expand <code>Podfile.lock</code> snippet</summary>
<br>

```yml

Replace this line with the contents of your Podfile.lock!

```

</details>

**Comment by paulb777** [comment]
This may be https://github.com/google/gtm-session-fetcher/pull/324

Does updating GTMSessionFetcher to at least version 2.2.0 help? 

In Xcode, do File -> Packages -> Update to Latest Versions

**Comment by ian-sayles** [comment]
Updated to latest version, 2.3.0 and still have same warning when I upload a file.

@@ -113,8 +113,8 @@
      "kind" : "remoteSourceControl",
      "location" : "https://github.com/google/gtm-session-fetcher.git",
      "state" : {
        "revision" : "d4289da23e978f37c344ea6a386e5546e2466294",
        "version" : "2.1.0"
        "revision" : "5ccda3981422a84186387dbb763ba739178b529c",
        "version" : "2.3.0"
      }
    },
    {

**Comment by nsarno-lucid** [comment]
I am getting the following data race condition:

```
==================
WARNING: ThreadSanitizer: data race (pid=805)
  Write of size 8 at 0x7b5400075db0 by thread T9 (mutexes: write M983609169334983936):
    #0 -[GTMSessionFetcher URLSession:dataTask:didReceiveData:] <null>:2 (APPNAME:x86_64+0x1036874db)
    #1 -[GTMSessionFetcherSessionDelegateDispatcher URLSession:dataTask:didReceiveData:] <null>:2 (APPNAME:x86_64:x86_64+0x1036bc59f)
    #2 <null> <null>:2 (CFNetwork:x86_64+0x102b6)
    #3 _dispatch_client_callout <null>:2 (libdispatch.dylib:x86_64+0x4f5a)

  Previous read of size 8 at 0x7b5400075db0 by main thread (mutexes: write M513410362349737768):
    #0 __56-[GTMSessionFetcher URLSession:dataTask:didReceiveData:]_block_invoke_2 <null>:2 (APPNAME:x86_64:x86_64+0x10368793d)
    #1 __66-[GTMSessionFetcher invokeOnCallbackQueue:afterUserStopped:block:]_block_invoke <null>:2 (APPNAME:x86_64:x86_64+0x103684f1d)
    #2 __wrap_dispatch_group_async_block_invoke <null>:2 (libclang_rt.tsan_iossim_dynamic.dylib:x86_64+0x806be)
    #3 _dispatch_client_callout <null>:2 (libdispatch.dylib:x86_64+0x4f5a)
    #4 <null> <null> (0x0001109c52bf)

  Location is heap block of size 592 at 0x7b5400075d00 allocated by thread T7:
    #0 calloc <null>:2 (libclang_rt.tsan_iossim_dynamic.dylib:x86_64+0x55614)
    #1 _objc_rootAllocWithZone <null>:2 (libobjc.A.dylib:x86_64+0x186e5)
    #2 -[GTMSessionFetcherService fetcherWithRequest:] <null>:2 (APPNAME:x86_64:x86_64+0x1036ac772)
    #3 closure #1 in StorageDownloadTask.enqueueImplementation(resumeWith:) <null>:2 (APPNAME:x86_64:x86_64+0x103621e7d)
    #4 partial apply for closure #1 in StorageDownloadTask.enqueueImplementation(resumeWith:) <null>:2 (APPNAME:x86_64:x86_64+0x103622985)
    #5 thunk for @escaping @callee_guaranteed () -> () <null>:2 (APPNAME:x86_64:x86_64+0x1035ea7c2)
    #6 __tsan::invoke_and_release_block(void*) <null>:2 (libclang_rt.tsan_iossim_dynamic.dylib:x86_64+0x7f3eb)
    #7 _dispatch_client_callout <null>:2 (libdispatch.dylib:x86_64+0x4f5a)

  Mutex M983609169334983936 is already destroyed.

  Mutex M513410362349737768 is already destroyed.

  Thread T9 (tid=22736255, running) is a GCD worker thread

  Thread T7 (tid=22736253, running) is a GCD worker thread

SUMMARY: ThreadSanitizer: data race (Lucid Dev:x86_64+0x1036874db) in -[GTMSessionFetcher URLSession:dataTask:didReceiveData:]+0x67b
==================
```

When calling:

```
private func getData(at reference: StorageReference, completion: @escaping (_ data: Data?, _ error: Error?) -> Void) {
	reference.getData(maxSize: 30 * 1024 * 1024, completion: completion)
}
```

For reference:

<img width="1127" alt="Screen Shot 2022-11-10 at 1 23 54 PM" src="https://user-images.githubusercontent.com/106601664/201176226-f726d775-6d55-4296-844d-9b4e7b21f141.png">


I found this thread, and updated "Update to Latest Versions" (per above comments) and now I'm getting the race condition **and** the priority inversion issue. Before updating, I was on GTMSessionFetcher either 2.1.0 or 2.2.0 and the priority inversion was not happening... and I don't know how to lower it back down now.

```
Thread Performance Checker: Thread running at QOS_CLASS_USER_INTERACTIVE waiting on a lower QoS thread running at QOS_CLASS_BACKGROUND. Investigate ways to avoid priority inversions
```

I see that there's a new version 3.0.0 of GTMSessionFetcher (https://github.com/google/gtm-session-fetcher/releases/tag/v3.0.0). I'm not sure if this would help, but the Firebase Package.swift is limited to < 3.0.0

```
.package(
      name: "GTMSessionFetcher",
      url: "https://github.com/google/gtm-session-fetcher.git",
      "2.1.0" ..< "3.0.0"
    ),

```

**Comment by thomasvl** [comment]
> Thanks for checking. 3.0.0 is unlikely to help since it's exactly the same as 2.3.0 except for the minimum iOS version specification.

This isn't correct; 3.0.0 is the same as 2.2.0.

But 2.2.0 broke semver because of the minos version bump; and the bump was needed for https://github.com/google/gtm-session-fetcher/pull/324, so 2.3.0 doesn't have that fix, 3.0 does.

**Comment by nsarno-lucid** [comment]
@paulb777 @thomasvl I just upgraded from `Firebase/10.1.0` to `Firebase/master` , which is running `GTMSessionFetcher 3.0.0` and the priority inversion is gone. My original race condition still exists though.

**Comment by thomasvl** [comment]
A data race would be a different issue from what this issue was opened for.

**Comment by paulb777** [comment]
I'm going to mark the priority inversion issue fixed with the upcoming Firebase 10.2 release.

Would you create another issue for the data race problem, ideally with more repro information.

Thanks for raising these!
